src/handshake/server.rs

```diff
@@ -83,15 +87,21 @@ impl<'a, T: AsyncRead + AsyncWrite + Unpin> Server<'a, T> {
     }
 
     /// Await an incoming client handshake request.
-    pub async fn receive_request(&mut self) -> Result<ClientRequest<'a>, Error> {
+    pub async fn receive_request<'r>(&'r mut self) -> Result<ClientRequest<'r>, Error> {
         self.buffer.clear();
+
         loop {
             crate::read(&mut self.socket, &mut self.buffer, BLOCK_SIZE).await?;
-            if let Parsing::Done { value, offset } = self.decode_request()? {
-                self.buffer.advance(offset);
```

Comment:
This mutable borrow of `self.buffer` prevented us from borrowing stuff in return value.